net: l2: ieee802154: improved test coverage

fgrandel opened the pull request:
This PR is the first of several PRs that will develop the IEEE 802.15.4 stack in direction of TSCH support, see the corresponding [RFC](https://github.com/zephyrproject-rtos/zephyr/issues/50336) that lays out the overall implementation approach.

Supporting IEEE 802.15.4-2020 features in general and TSCH in particular requires substantial changes to the stack in order to align it with IEEE 802.15.4-2020 concepts, especially in the area of MHR parsing, support of arbitrary optional information elements (IEs), standards confirming configuration of MAC/PHY PIB attributes and incoming and outgoing security procedures.

The internal API of the IEEE 802.15.4 stack will have to be made more extensible and configurable to support these features. A preview of further changes to come is given in the [open development branch](https://github.com/zephyrproject-rtos/zephyr/compare/main...fgrandel:zephyr:feat/ieee802154-tsch-ie-support) to motivate why such preparative work is required.

**To ensure that none of these upcoming changes will break backwards compatibility, broad test coverage is introduced in this first PR.**

Commit:
net: l2: ieee802154: test: raw packet readability

Improves inline documentation of binary IEEE 802.15.4 frames
to ease future maintenance. In this case the frame required
for testing raw sockets.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
net: l2: ieee802154: test: extract security set up/tear down

Security set-up and clean-up is extracted into functions
to enable re-use by test cases introduced in later commits.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
net: l2: ieee802154: test: cover dgram reception

Additional test case that covers reception of AF_SOCKET/SOCK_DGRAM
packages via socket API.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
net: l2: ieee802154: test: prepare expectation struct enlargement

This is a purely syntactical change that makes test case
expectations configuration more readable.

This is required as later commits will introduce further
expectations.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
net: l2: ieee802154: test: cover ACK procedure

This change introduces test coverage for the IEEE 802.15.4
acknowledgement procedure.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
net: l2: ieee802154: test: cover raw reception

Additional test case that covers AF_SOCKET/SOCK_RAW package
reception.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
net: l2: ieee802154: test: extract data pkt with ar

Additional test cases introduced in later commits will re-use
an existing test frame that requests acknowledgement. It is
therefore extracted into a function.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
net: l2: ieee802154: test: /w and /wo sockets

This change splits the IEEE 802.15.4 test into two separate
test profiles, one with and one without sockets enabled to ensure
that both configurations work correctly.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
net: l2: ieee802154: test: fix resource mgmt

This change fixes some minor resource leaks in the test suite.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
net: l2: ieee802154: test: simlify sockaddr config

This refactoring dries up duplicate code required to configure
the socket address structure used in tests.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
net: l2: ieee802154: test: improved pkt doc

Introduces inline documentation of binary test frames to ease
future maintenance.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
net: l2: ieee802154: test: sequence number coverage

IEEE 802.15.4-2020 introduces a sequence number suppression feature.

This change covers sequence number handling to ensure that future
changes will not break sequence number handling when introducing
the sequence number suppression feature.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
net: l2: ieee802154: expose implicit frame assumption

The current IEEE 802.15.4 stack assumes that drivers will
not provide more than one fragment in incoming packages.

This change exposes and enforces the pre-existing assumption
made by the implementation.

So far this is not a limiting restriction as MPDUs with more
than 127 bytes are not supported yet. It will probably have to
be changed as soon as larger payloads (as allowed by PHYs
introduced more recently into the spec) might want to allocate
smaller fragments.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

fgrandel commented:
@nandojve Heads up to you as you might review this PR as well.

rlubos commented on tests/net/ieee802154/l2/src/ieee802154_test.c:
```diff
@@ -983,6 +987,7 @@ ZTEST(ieee802154_l2, test_parsing_ack_pkt)
 	zassert_true(ret, "ACK parsed");
 }
 
+#ifdef CONFIG_NET_SOCKETS
```
I wonder whether it'd make sense to define a separate `ZTEST_SUITE()` for raw socket tests?

fgrandel replied:
Gotcha, will do.

fgrandel replied:
Fixed.

rlubos reviewed (commented):
Just some nits, but overall it looks quite reasonable to me.

fgrandel commented on tests/net/ieee802154/l2/src/ieee802154_test.c:
```diff
@@ -177,6 +177,42 @@ static int tear_down_short_addr(struct net_if *iface, struct ieee802154_context
+	struct ieee802154_security_params params;
+
+	if (security_level == IEEE802154_SECURITY_LEVEL_NONE)
+		return true;
```
Fixed

fgrandel commented on tests/net/ieee802154/l2/src/ieee802154_test.c:
```diff
@@ -177,6 +177,42 @@ static int tear_down_short_addr(struct net_if *iface, struct ieee802154_context
+{
+	struct ieee802154_context *ctx = net_if_l2_data(iface);
+
+	if (ctx->sec_ctx.level == IEEE802154_SECURITY_LEVEL_NONE)
```
Fixed

nandojve commented on tests/net/ieee802154/l2/src/ieee802154_test.c:
```diff
@@ -177,6 +177,44 @@ static int tear_down_short_addr(struct net_if *iface, struct ieee802154_context
+
+static void tear_down_security(void)
+{
+	struct ieee802154_context *ctx = net_if_l2_data(iface);
+
+	if (ctx->sec_ctx.level == IEEE802154_SECURITY_LEVEL_NONE) {
+		return;
+	}
+
+	cipher_free_session(ctx->sec_ctx.enc.device, &ctx->sec_ctx.enc);
+	cipher_free_session(ctx->sec_ctx.dec.device, &ctx->sec_ctx.dec);
+	ctx->sec_ctx.level = IEEE802154_SECURITY_LEVEL_NONE;
```
Did you know why it is necessary clean-up using cipher* functions instead make a request passing security_level as IEEE802154_SECURITY_LEVEL_NONE ? This seems to be a little hack to me.

fgrandel replied:
This is a very good question. The fact that the security context is not cleaned up inside the mgmt-stack before starting a new one is even a resource leak. Fixed.

nandojve reviewed (dismissed):
Hi @fgrandel ,

I only noted that when you wrote the tear down seems that expose something little odd to me.

Thank you for heads up and these changes.

Commit:
net: l2: ieee802154: security: properly clean up previous session

When changing IEEE 802.15.4 security settings or setting security
to 'none' then the previous session must be cleaned up to avoid
resource leaks.

This change introduces proper clean-up of the security session.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

jukkar reviewed (approved):
LGTM
